examples/cpp/helloworld/greeter_async_bidi_client.cc

```diff
@@ -0,0 +1,213 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+
+#include <grpc++/grpc++.h>
+
+#include "hellostreamingworld.grpc.pb.h"
+
+using grpc::Channel;
+using grpc::ClientAsyncReaderWriter;
+using grpc::ClientContext;
+using grpc::CompletionQueue;
+using grpc::Status;
+using hellostreamingworld::HelloRequest;
+using hellostreamingworld::HelloReply;
+using hellostreamingworld::MultiGreeter;
+
+// NOTE: This is a complex example for an asynchronous, bidirectional streaming
+// client. For a simpler example, start with the
+// greeter_client/greeter_async_client first.
+class AsyncBidiGreeterClient {
+  enum class Type {
+    READ = 1,
+    WRITE = 2,
+    CONNECT = 3,
+    WRITES_DONE = 4,
+    FINISH = 5
+  };
+
+ public:
+  explicit AsyncBidiGreeterClient(std::shared_ptr<Channel> channel)
+      : stub_(MultiGreeter::NewStub(channel)) {
+    grpc_thread_.reset(
+        new std::thread(std::bind(&AsyncBidiGreeterClient::GrpcThread, this)));
+    stream_ = stub_->AsyncSayHello(&context_, &cq_,
+                                   reinterpret_cast<void*>(Type::CONNECT));
+  }
+
+  // Similar to the async hello example in greeter_async_client but does not
+  // wait for the response. Instead queues up a tag in the completion queue
+  // that is notified when the server responds back (or when the stream is
+  // closed). Returns false when the stream is requested to be closed.
+  bool AsyncSayHello(const std::string& user) {
+    if (user == "quit") {
+      stream_->WritesDone(reinterpret_cast<void*>(Type::WRITES_DONE));
+      return false;
+    }
+
+    // Data we are sending to the server.
+    HelloRequest request;
+    request.set_name(user);
+
+    // This is important: You can have at most one write or at most one read
+    // at any given time. The throttling is performed by gRPC completion
+    // queue. If you queue more than one write/read, the stream will crash.
+    // Because this stream is bidirectional, you *can* have a single read
+    // and a single write request queued for the same stream. Writes and reads
+    // are independent of each other in terms of ordering/delivery.
+    std::cout << " ** Sending request: " << user << std::endl;
+    stream_->Write(request, reinterpret_cast<void*>(Type::WRITE));
+    return true;
+  }
+
+  ~AsyncBidiGreeterClient() {
+    std::cout << "Shutting down client...." << std::endl;
+    grpc::Status status;
+    cq_.Shutdown();
+    grpc_thread_->join();
```

Comment:
So this is tricky. I think it will work in this case, but not necessarily in general. In particular, if the thread code calls any virtual function of this class, you'll have the problem of having a vtbl reference concurrent with the destructor. I like ending multithreading before getting to the destructor. In this case, it doesn't matter probably since there are no virtual functions; maybe declare the class `final` to leave no doubt.